README.md
# syphon

A new Flutter project.

## Getting Started

This project is a starting point for a Flutter application.

A few resources to get you started if this is your first Flutter project:

- [Lab: Write your first Flutter app](https://flutter.dev/docs/get-started/codelab)
- [Cookbook: Useful Flutter samples](https://flutter.dev/docs/cookbook)

For help getting started with Flutter, view our
[online documentation](https://flutter.dev/docs), which offers tutorials,
samples, guidance on mobile development, and a full API reference.


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.2.7](https://github.com/syphon-org/syphon/releases/tag/0.2.6) - 2021-12-27

### Added
- removes exif information for image uploads

### Fixed
- fixed sending images from the quick select menu
- fixed inviting users to chats
- fixed removing deleted images correctly from UI 


## [0.2.6](https://github.com/syphon-org/syphon/releases/tag/0.2.6) - 2021-12-27

### Fixed
- screen lock regression on confirming after toggled on

## [0.2.5](https://github.com/syphon-org/syphon/releases/tag/0.2.5) - 2021-12-24

### Fixed
- issue with chat colors and message details screens

## [0.2.4](https://github.com/syphon-org/syphon/releases/tag/0.2.4) - 2021-12-24

### Added 
- Screen Lock
- Pin Protected Encrypted Storage
- Improved Edit Message UI
- Improved Chat UI spacing
- UserID Quick Copy

### Fixed
- major performance improvements and optimizations
- fixed editing encrypted messages 
- fixed deleting encrypted messages
- fixed read receipts locally if toggled on
- fixed defaulting to device language
- fixed editing messages higher in the chat
- fixed opening links in Android 11
- fixed showing more icon in user previews
- fixed input text border regression

### Changed

## [0.1.6](https://github.com/syphon-org/syphon/releases/tag/0.1.5) - 20XX-XX-XX

### Added 
- cold storage backups of all essential cache stores (redundant / failover local backups)

### Fixed
- Fixed issue where avatar urls would load indefinitely if not found
- Fixed issue where login/signup information could get saved in the encrypted cache

### Changed

## [0.1.5](https://github.com/syphon-org/syphon/releases/tag/0.1.5) - 2020-12-15

### Added 
- Block users (removes direct chats and hides messages)

### Fixed
- Lots of improvements in regards to storing data and initial sign in
- Overhauled caching again for performance and desktop client support

### Changed 
- Old cache management (hive) was removed, may need to sign out/sign in

## [0.1.4](https://github.com/syphon-org/syphon/releases/tag/0.1.4) - 2020-10-28

### Added
- Invite Friends from Chat Menu (if allowed to)
- Avatars can be set to square shape in Theme settings
- Much better Local Notification handling with room names

### Fixed
- Overhauled caching to solve lag on home screen
- Lazy loading data on the initial sync
- Uses can no longer send empty messages
- Users of homeservers with a forward slash "/" in their m.homeserver config can login
- Issues with room names appearing as the current user
- Issues with room avatars appearing as joined users
- Password restrictions on login were tied to signup password restrictions
- Intermittent syncing issues when not opening the app for a while
- Smaller bug fixes and performance improvements

## [0.1.3](https://github.com/syphon-org/syphon/releases/tag/0.1.3-4) - 2020-09-18

### Added
- New Icon!
- Native splash screens
- Better Performance on first load
- Toggle 24 Hour Time Format

### Fixed
- E2E had cross client pre-key sharing issues. Element to Syphon messages work again.
- Better anonymous, human readable device IDs that are still unique to your device.
- General bug fixes and performance improvements
- Other issues regarding FDroid deployments and auto updates


## [0.1.2](https://github.com/syphon-org/syphon/releases/tag/0.1.2) - 2020-08-11

### Added 
- Private group E2EE encryption*
- Create public and private groups - w/ photos
- User Profiles (both quick view and fullscreen
- Invite users to any joined room
- mark all messages read (locally or remotely)
- tons of code refactoring to help dev and performance
- a lot more (new fonts, colors, etc)

### Fixed
- Confirmation Alert Improvements
- Standardized view mounting and unmounting
- Disabled all non-available features in menus
- New Action Ring Icons

### Removed
- Dead clickable links in menus

## [0.0.23](https://github.com/syphon-org/syphon/releases/tag/0.0.23) - 2020-07-20
### Added
- hot fixes for FDroid and iOS publishing

## [0.0.21](https://github.com/syphon-org/syphon/releases/tag/0.0.21) - 2020-07-20
### Added
- opt-in read receipts
- signup email input and verification (working with matrix.org again)
- signup/login bug fixes, standardizing secure text inputs
- global error alerts and confirmations will show up
- better small screen UI scaling
- lots of code cleanup and refactoring


lib/views/login.dart
import 'package:Tether/domain/user/actions.dart';
import 'package:Tether/domain/user/selectors.dart';
import 'package:Tether/global/strings.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';

import 'package:redux/redux.dart';
import 'package:flutter_redux/flutter_redux.dart';

// Domain
import 'package:Tether/domain/index.dart';
import 'package:Tether/domain/settings/actions.dart';

// Styling
import 'package:touchable_opacity/touchable_opacity.dart';
import 'package:Tether/global/dimensions.dart';
import 'package:Tether/global/behaviors.dart';

// Assets
import 'package:Tether/global/assets.dart';

class Login extends StatefulWidget {
  final Store<AppState> store;
  const Login({Key key, this.store}) : super(key: key);

  @override
  LoginState createState() => LoginState(store: this.store);
}

class LoginState extends State<Login> {
  final GlobalKey<ScaffoldState> loginScaffold = GlobalKey<ScaffoldState>();
  final usernameController = TextEditingController();
  final passwordController = TextEditingController();
  final passwordFocus = FocusNode();

  final Store<AppState> store;

  LoginState({Key key, this.store});

  @override
  void initState() {
    super.initState();

    SchedulerBinding.instance.addPostFrameCallback((_) {
      onMounted();
    });
  }

  @override
  void dispose() {
    usernameController.dispose();
    passwordController.dispose();
    passwordFocus.dispose();
    super.dispose();
  }

  @protected
  void onMounted() {
    // Init alerts listener
    store.state.alertsStore.onAlertsChanged.listen((alert) {
      var color;

      switch (alert.type) {
        case 'warning':
          color = Colors.red;
          break;
        case 'error':
          color = Colors.red;
          break;
        case 'info':
        default:
          color = Colors.grey;
      }

      loginScaffold.currentState.showSnackBar(SnackBar(
        backgroundColor: color,
        content: Text(alert.message),
        duration: alert.duration,
        action: SnackBarAction(
          label: 'Dismiss',
          textColor: Colors.white,
          onPressed: () {
            loginScaffold.currentState.removeCurrentSnackBar();
          },
        ),
      ));
    });
  }

  void handleSubmitted(String value) {
    FocusScope.of(context).requestFocus(passwordFocus);
  }

  @override
  Widget build(BuildContext context) {
    double width = MediaQuery.of(context).size.width;
    double height = MediaQuery.of(context).size.height;
    final double defaultWidgetScaling = width * 0.725;

    return Scaffold(
      key: loginScaffold,
      body: ScrollConfiguration(
        behavior: DefaultScrollBehavior(),
        child: SingleChildScrollView(
          // Use a container of the same height and width
          // to flex dynamically but within a single child scroll
          child: StoreConnector<AppState, Store<AppState>>(
            converter: (store) => store,
            builder: (context, store) => Container(
              height: height,
              constraints: BoxConstraints(
                maxHeight: MAX_WIDGET_HEIGHT,
              ),
              child: Flex(
                direction: Axis.vertical,
                mainAxisAlignment: MainAxisAlignment.center,
                children: <Widget>[
                  Flexible(
                    flex: 4,
                    child: Flex(
                      direction: Axis.vertical,
                      mainAxisAlignment: MainAxisAlignment.end,
                      children: <Widget>[
                        TouchableOpacity(
                          onTap: () {
                            store.dispatch(
                              incrementTheme(),
                            );
                          },
                          child: Image(
                            width: width * 0.35,
                            height: width * 0.35,
                            image: AssetImage(TETHER_ICON_PNG),
                          ),
                        ),
                      ],
                    ),
                  ),
                  Flexible(
                    flex: 2,
                    child: Flex(
                        direction: Axis.vertical,
                        mainAxisAlignment: MainAxisAlignment.center,
                        children: [
                          FittedBox(
                            fit: BoxFit.fitWidth,
                            child: Text(
                              LOGIN_TITLE,
                              textAlign: TextAlign.center,
                              style: Theme.of(context).textTheme.headline4,
                            ),
                          ),
                        ]),
                  ),
                  Flexible(
                    flex: 3,
                    fit: FlexFit.loose,
                    child: Flex(
                        direction: Axis.vertical,
                        mainAxisAlignment: MainAxisAlignment.end,
                        children: [
                          Container(
                            width: defaultWidgetScaling,
                            height: DEFAULT_INPUT_HEIGHT,
                            margin: const EdgeInsets.symmetric(
                              vertical: 8,
                            ),
                            constraints: BoxConstraints(
                              minWidth: MIN_INPUT_WIDTH,
                              maxWidth: MAX_INPUT_WIDTH,
                            ),
                            child: TextField(
                              controller: usernameController,
                              onSubmitted: handleSubmitted,
                              onChanged: (username) {
                                // Trim value for UI
                                usernameController.value = TextEditingValue(
                                  text: username.trim(),
                                  selection: TextSelection.fromPosition(
                                    TextPosition(
                                      offset: username.trim().length,
                                    ),
                                  ),
                                );
                                // If user enters full username, make sure to set homeserver
                                if (username.contains(':')) {
                                  final alias = username.trim().split(':');
                                  store.dispatch(setUsername(
                                    username: alias[0],
                                  ));
                                  store.dispatch(setHomeserver(
                                    homeserver: alias[1],
                                  ));
                                } else {
                                  store.dispatch(setUsername(
                                    username: username.trim(),
                                  ));
                                }
                              },
                              decoration: InputDecoration(
                                contentPadding: EdgeInsets.only(
                                  left: 20,
                                  top: 32,
                                ),
                                suffixIcon: IconButton(
                                  icon: Icon(Icons.help_outline),
                                  tooltip: SELECT_USERNAME_TITLE,
                                  onPressed: () {
                                    Navigator.pushNamed(
                                      context,
                                      '/search_home',
                                    );
                                  },
                                ),
                                border: OutlineInputBorder(
                                  borderRadius: BorderRadius.circular(28.0),
                                ),
                                hintText: formatUsernameHint(
                                  store.state.userStore.homeserver,
                                ),
                                labelText: 'username',
                              ),
                            ),
                          ),
                          Container(
                            width: defaultWidgetScaling,
                            height: DEFAULT_INPUT_HEIGHT,
                            margin: const EdgeInsets.symmetric(
                              vertical: 8,
                            ),
                            constraints: BoxConstraints(
                              minWidth: MIN_INPUT_WIDTH,
                              maxWidth: MAX_INPUT_WIDTH,
                            ),
                            child: TextField(
                              focusNode: passwordFocus,
                              onChanged: (password) {
                                store.dispatch(setPassword(password: password));
                              },
                              obscureText: true,
                              decoration: InputDecoration(
                                contentPadding: EdgeInsets.only(
                                  left: 20,
                                  top: 32,
                                  bottom: 32,
                                ),
                                border: OutlineInputBorder(
                                  borderRadius: BorderRadius.circular(30.0),
                                ),
                                labelText: 'password',
                              ),
                            ),
                          ),
                        ]),
                  ),
                  Container(
                    width: defaultWidgetScaling,
                    height: DEFAULT_BUTTON_HEIGHT,
                    margin: const EdgeInsets.only(
                      top: 24,
                    ),
                    constraints: BoxConstraints(
                      minWidth: 256,
                      maxWidth: 336,
                    ),
                    child: FlatButton(
                      disabledColor: Colors.grey,
                      disabledTextColor: Colors.grey[300],
                      onPressed: isLoginAttemptable(store.state)
                          ? () {
                              store.dispatch(loginUser());
                            }
                          : null,
                      color: Theme.of(context).primaryColor,
                      shape: RoundedRectangleBorder(
                        borderRadius: BorderRadius.circular(28.0),
                      ),
                      child: isAuthLoading(store.state)
                          ? Container(
                              constraints: BoxConstraints(
                                maxHeight: 28,
                                maxWidth: 28,
                              ),
                              child: CircularProgressIndicator(
                                strokeWidth: 2,
                                backgroundColor: Colors.white,
                                valueColor: AlwaysStoppedAnimation<Color>(
                                  Colors.grey,
                                ),
                              ),
                            )
                          : Text(
                              LOGIN_BUTTON_TEXT,
                              style: TextStyle(
                                fontSize: 20,
                                color: Colors.white,
                              ),
                            ),
                    ),
                  ),
                  Container(
                    height: DEFAULT_INPUT_HEIGHT,
                    constraints: BoxConstraints(
                      minHeight: DEFAULT_BUTTON_HEIGHT,
                    ),
                    margin: const EdgeInsets.symmetric(
                      horizontal: 8,
                      vertical: 16,
                    ),
                    child: TouchableOpacity(
                      activeOpacity: 0.4,
                      onTap: () => Navigator.pushNamed(
                        context,
                        '/signup',
                      ),
                      child: Row(
                        mainAxisAlignment: MainAxisAlignment.center,
                        children: <Widget>[
                          Text(
                            CREATE_USER_TEXT,
                            textAlign: TextAlign.center,
                            style: TextStyle(
                              fontSize: 18,
                              fontWeight: FontWeight.w100,
                            ),
                          ),
                          Container(
                            padding: const EdgeInsets.only(left: 4),
                            child: Text(
                              CREATE_USER_TEXT_ACTION,
                              textAlign: TextAlign.center,
                              style: TextStyle(
                                fontSize: 18,
                                fontWeight: FontWeight.w100,
                                color: Theme.of(context).primaryColor,
                                decoration: TextDecoration.underline,
                              ),
                            ),
                          ),
                        ],
                      ),
                    ),
                  ),
                ],
              ),
            ),
          ),
        ),
      ),
    );
  }
}
